staging/src/k8s.io/client-go/tools/cache/shared_informer_health.go

```diff
@@ -0,0 +1,182 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cache
+
+import (
+	"fmt"
+	"net/http"
+	"sync/atomic"
+	"time"
+
+	"k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/apimachinery/pkg/util/wait"
+)
+
+func newDefaultListWatchLockFreeHealth(name string, errorReceiverCh <-chan error) *listWatchLockFreeHealth {
+	lwlf := &listWatchLockFreeHealth{
+		name:                name,
+		minTimeToFail:       time.Duration(5 * time.Second),
+		minConsistentErrors: 10,
+		consistentErrorsCh:  errorReceiverCh,
+		errCircularBuffer:   atomic.Value{},
+	}
+
+	lwlf.errCircularBuffer.Store(newErrBuffer(5))
+	return lwlf
+}
+
+type listWatchLockFreeHealth struct {
+	name         string
+	startingTime int64
+
+	// minTimeToFail minimum time duration after which we start to fail
+	minTimeToFail time.Duration
+	// minConsistentError minimum number of errors before we start to fail
+	minConsistentErrors int32
+	// consistentErrors holds the number of errors seen so far
+	consistentErrors int32
+
+	// consistentErrorsCh channel on which we receive List&Watch related errors
+	// it is meant to represent a gauge, that is a number that can go up and down
+	consistentErrorsCh <-chan error
+
+	errCircularBuffer atomic.Value
+}
+
+func (lwlf *listWatchLockFreeHealth) run(stopCh <-chan struct{}) {
+	go wait.Until(func() {
+		lwlf.runInternal(stopCh)
+	}, 1*time.Minute, stopCh)
+}
+
+func (lwlf *listWatchLockFreeHealth) runInternal(stopCh <-chan struct{}) {
+	for {
+		select {
+		case <-stopCh:
+			return
+		case errRcv, ok := <-lwlf.consistentErrorsCh:
+			if !ok {
+				break
+			}
+			seenErrors := atomic.LoadInt32(&lwlf.consistentErrors)
+			if errRcv == nil {
```

Comment:
comment explaining that means that a reflector has transitioned to a non-failing state